io.openems.edge.evcs.alfen/src/io/openems/edge/evcs/alfen/AlfenImpl.java

```diff
@@ -0,0 +1,340 @@
+import io.openems.common.exceptions.OpenemsError.OpenemsNamedException;
+import io.openems.common.exceptions.OpenemsException;
+import io.openems.edge.bridge.modbus.api.AbstractOpenemsModbusComponent;
+import io.openems.edge.bridge.modbus.api.BridgeModbus;
+import io.openems.edge.bridge.modbus.api.ModbusComponent;
+import io.openems.edge.bridge.modbus.api.ModbusProtocol;
+import io.openems.edge.bridge.modbus.api.element.FloatDoublewordElement;
+import io.openems.edge.bridge.modbus.api.element.FloatQuadruplewordElement;
+import io.openems.edge.bridge.modbus.api.element.StringWordElement;
+import io.openems.edge.bridge.modbus.api.element.UnsignedDoublewordElement;
+import io.openems.edge.bridge.modbus.api.element.UnsignedQuadruplewordElement;
+import io.openems.edge.bridge.modbus.api.element.UnsignedWordElement;
+import io.openems.edge.bridge.modbus.api.task.FC16WriteRegistersTask;
+import io.openems.edge.bridge.modbus.api.task.FC3ReadRegistersTask;
+import io.openems.edge.common.channel.value.Value;
+import io.openems.edge.common.component.OpenemsComponent;
+import io.openems.edge.common.event.EdgeEventConstants;
+import io.openems.edge.common.taskmanager.Priority;
+import io.openems.edge.evcs.api.Evcs;
+import io.openems.edge.evcs.api.EvcsPower;
+import io.openems.edge.evcs.api.ManagedEvcs;
+import io.openems.edge.evcs.api.Status;
+
+@Designate(ocd = Config.class, factory = true)
+@Component(//
+		name = "Evcs.Alfen", //
+		immediate = true, //
+		configurationPolicy = ConfigurationPolicy.REQUIRE, //
+		property = { //
+				EventConstants.EVENT_TOPIC + "=" + EdgeEventConstants.TOPIC_CYCLE_BEFORE_PROCESS_IMAGE, //
+				EventConstants.EVENT_TOPIC + "=" + EdgeEventConstants.TOPIC_CYCLE_AFTER_PROCESS_IMAGE //
+		} //
+)
+public class AlfenImpl extends AbstractOpenemsModbusComponent
+		implements Alfen, Evcs, ManagedEvcs, ModbusComponent, EventHandler, OpenemsComponent {
+
+	private final Logger log = LoggerFactory.getLogger(AlfenImpl.class);
+
+	private static final float DETECT_PHASE_ACTIVITY = 400; // milliAmpere
+
+	private Long energyAtSessionStart = 0L;
+
+	@Reference
+	protected ConfigurationAdmin cm;
+
+	@Reference
+	private EvcsPower evcsPower;
+
+	@Reference(policy = ReferencePolicy.STATIC, policyOption = ReferencePolicyOption.GREEDY, cardinality = ReferenceCardinality.MANDATORY)
+	protected void setModbus(BridgeModbus modbus) {
+		super.setModbus(modbus);
+	}
+
+	private int phasePattern = 0;
+
+	public AlfenImpl() {
+		super(//
+				OpenemsComponent.ChannelId.values(), //
+				ModbusComponent.ChannelId.values(), //
+				Evcs.ChannelId.values(), //
+				ManagedEvcs.ChannelId.values(), //
+				Alfen.ChannelId.values() //
+		);
+	}
+
+	@Activate
+	void activate(ComponentContext context, Config config) throws OpenemsException {
+		if (super.activate(context, config.id(), config.alias(), config.enabled(), config.modbusUnitId(), this.cm,
+				"Modbus", config.modbus_id())) {
+			return;
+		}
+
+		this._setMaximumHardwarePower(22000);
```

Comment:
This values should be configurable. We should handle the configuration same as Keba implementation does - by using mA for min and max range of chargepoint. Note that this has to do with the UI Renault Zoe option, which reconfigures the chargepoint by setting a different min current (in mA)